referrer-policy/generic/common.js

```diff
@@ -111,23 +111,43 @@ function queryIframe(url, callback, referrer_policy) {
   window.addEventListener("message", listener);
 }
 
-function queryImage(url, callback, referrer_policy) {
-  // For images, we'll test both images in a top-level document as well as
-  // images in a `srcdoc` frame to ensure that the latter has the same referrer
-  // as the former.
-  var i = document.createElement('iframe');
-  i.srcdoc = "Hello, world.";
-  i.onload = function () {
+function queryImage(url, callback, attributes, referrerPolicy) {
+  // For images, we'll test:
+  // - images in a `srcdoc` frame to ensure that it uses the referrer
+  //   policy of its parent,
+  // - images in a top-level document,
+  // - and images in a `srcdoc` frame with its own referrer policy to
+  //   override its parent.
+  var iframeWithoutOwnPolicy = document.createElement('iframe');
+  iframeWithoutOwnPolicy.srcdoc = "Hello, world.";
+
+  iframeWithoutOwnPolicy.onload = function () {
     loadImageInWindow(url, function (img) {
       var srcdocData = decodeImageData(extractImageData(img));
-      loadImageInWindow(url, function (img) {
-        var topLevelData = decodeImageData(extractImageData(img));
-        assert_equals(srcdocData.referrer, topLevelData.referrer, "Referrer inside 'srcdoc' should be the same as embedder's referrer.");
-        callback(wrapResult(url, topLevelData), url);
-      }, referrer_policy, window);
-    }, referrer_policy, i.contentWindow);
+      var iframeWithOwnPolicy = document.createElement('iframe');
+      // Give a srcdoc iframe a referrer policy different from the top-level page's policy.
+      var iframePolicy = (referrerPolicy === "no-referrer") ? "unsafe-url" : "no-referrer";
+      iframeWithOwnPolicy.srcdoc = "<meta name='referrer' content='" + iframePolicy + "'>Hello world.";
+
+      iframeWithOwnPolicy.onload = function () {
+        var nextUrl = url + "&cache_destroyer2=" + (new Date()).getTime();
+        loadImageInWindow(nextUrl, function (img) {
+          assert_equals((iframePolicy === "no-referrer" ? undefined : document.location.href), decodeImageData(extractImageData(img)).headers.referer, "Referrer inside 'srcdoc' should use the iframe's policy if it has one");
+
+          nextUrl = url + "&cache_destroyer3=" + (new Date()).getTime();
+          loadImageInWindow(nextUrl, function (img) {
```

Comment:
Indeed, this is pretty tough to follow. Especially since we've pushed the frame in which we're working all the way to the end of the function.

Since none of the frames depends on the other, could you perhaps rearrange things like:

```
var iframeWithoutOwnPolicy = ...;
iframeWithoutOwnPolicy.onload = function () {
  loadImageInWindow({ .... }, iframeWithoutOwnPolicy.contentWindow);
};
var iframeWithOwnPolicy = ...;
iframeWithOwnPolicy.onload = function () {
  loadImageInWindow({ .... }, iframeWithOwnPolicy.contentWindow);
};
```

And gate the call to `callback()` on both completing? You could set a bool for both, and end both with `maybeCallCallback()` which would be something like `if (firstFrameIsDone && secondFrameIsDone) { callback(wrapResult(...), url)`?